File: 03/Perl/t3.pl
#!/usr/bin/perl 
#===============================================================================
#
#         FILE: t3.pl
#
#        USAGE: ./t3.pl  
#
#  DESCRIPTION: 
#
#      OPTIONS: ---
# REQUIREMENTS: ---
#         BUGS: ---
#        NOTES: ---
#       AUTHOR: YOUR NAME (), 
# ORGANIZATION: 
#      VERSION: 1.0
#      CREATED: 12/03/2019 09:48:28 AM
#     REVISION: ---
#===============================================================================

use strict;
use warnings;
use Data::Dumper;
use 5.020;

my %grid;
my $posx=0;
my $posy=0;
my %seen;

sub process_step {
                $grid{$posx}{$posy}++ unless $seen{$posx}{$posy};
                $seen{$posx}{$posy}=1;

                if ($grid{$posx}{$posy} > 1) {
                    my $distance = abs($posx) + abs($posy);
                    say "intersection $posx $posy distance $distance";
                }
}

open my $file, '<', '../input' or die 'cannot open input';



while (<$file>) {
    $posx=0;
    $posy=0;
    %seen = ();

    chomp;

    my @arr = split /,/, $_;

    for (@arr) {
        my ($direction, $count) = $_ =~ /(.)(.*)/msx;

        if ($direction eq 'R') {
            for (1..$count) {
                $posx++;
                &process_step;
            }
        }

        if ($direction eq 'L') {
            for (1..$count) {
                $posx--;
                &process_step;
            }
        }

        if ($direction eq 'U') {
            for (1..$count) {
                $posy++;
                &process_step;
            }
        }

        if ($direction eq 'D') {
            for (1..$count) {
                $posy--;
                &process_step;
            }
        }

    }

}

File: 03/Python/t3.py
#! python3_

from pprint import pprint
import re

grid={}

def process_step(posx,posy) :
                if (posx,posy) not in seen:
                    if (posx,posy) not in grid:
                        grid[(posx,posy)] = 1
                    else:
                        grid[(posx,posy)] += 1
                        distance = abs(posx) + abs(posy)
                        print("intersection "+str(posx)+" "+str(posy)+" distance "+str(distance))

                seen[(posx,posy)]=1

file = open("../input","r")

for line in file:
        arr = line.strip().split(',')
        posx=0
        posy=0
        seen={}
        
        for item in arr:
            match=re.match(r"(.)(.*)", item)
            direction=match.group(1)
            count=match.group(2)
   
            if (direction == 'R') :
                for i in range(1,int(count)+1):
                    posx += 1
                    process_step(posx,posy)
            
            if (direction == 'L') :
                for i in range(1,int(count)+1):
                    posx -= 1
                    process_step(posx,posy)
 
            if (direction == 'U') :
                for i in range(1,int(count)+1):
                    posy += 1
                    process_step(posx,posy)


            if (direction == 'D') :
                for i in range(1,int(count)+1):
                    posy -= 1
                    process_step(posx,posy)


